File: backend/internal/services/product_service.go
package services

import (
	"context"
	"time"

	"github.com/google/uuid"
	"github.com/grawizah/backend/internal/interfaces"
	"github.com/grawizah/backend/internal/models"
)

type ProductService struct {
	productRepo interfaces.ProductRepository
}

func NewProductService(productRepo interfaces.ProductRepository) *ProductService {
	return &ProductService{
		productRepo: productRepo,
	}
}

func (s *ProductService) GetProducts(ctx context.Context, limit, offset int) ([]models.Product, error) {
	return s.productRepo.GetAll(limit, offset)
}

func (s *ProductService) GetProductByID(ctx context.Context, id string) (*models.Product, error) {
	return s.productRepo.GetByID(id)
}

func (s *ProductService) CreateProduct(ctx context.Context, product *models.Product) error {
	if product.ID == "" {
		product.ID = uuid.New().String()
	}
	now := time.Now()
	product.CreatedAt = &now
	product.UpdatedAt = &now

	// Calculate listing score before saving
	product.CalculateListingScore()
	return s.productRepo.Create(product)
}

func (s *ProductService) UpdateProduct(ctx context.Context, id string, product *models.Product) error {
	product.ID = id
	// Recalculate listing score
	product.CalculateListingScore()
	return s.productRepo.Update(product)
}

func (s *ProductService) DeleteProduct(ctx context.Context, id string) error {
	return s.productRepo.Delete(id)
}

func (s *ProductService) SearchProducts(ctx context.Context, query string, limit, offset int) ([]models.Product, error) {
	return s.productRepo.Search(query, limit, offset)
}

func (s *ProductService) IncrementViewCount(ctx context.Context, productID string) error {
	return s.productRepo.IncrementViewCount(productID)
}

func (s *ProductService) GetProductsByCategory(ctx context.Context, category string) ([]models.Product, error) {
	return s.productRepo.GetByCategory(category)
}

func (s *ProductService) GetProductsByCompany(ctx context.Context, companyID string) ([]models.Product, error) {
	return s.productRepo.GetByCompanyID(companyID)
}

File: frontend/src/services/ProductService.ts
import { BaseService } from './BaseService';
import { Product, ProductListingOptimization } from '@/types/product';
import { PaginatedResponse, PaginationParams } from '@/types';

/**
 * Product Service
 * Implements Inheritance (extends BaseService)
 */
export class ProductService extends BaseService {
  private readonly endpoint = '/api/products';

  async getProducts(params?: PaginationParams): Promise<PaginatedResponse<Product>> {
    return this.get<PaginatedResponse<Product>>(this.endpoint, { params });
  }

  async getProductById(id: string): Promise<Product> {
    return this.get<Product>(`${this.endpoint}/${id}`);
  }

  async searchProducts(query: string, filters?: any): Promise<Product[]> {
    return this.post<Product[]>(`${this.endpoint}/search`, { query, filters });
  }

  async getProductsByCategory(category: string): Promise<Product[]> {
    return this.get<Product[]>(`${this.endpoint}/category/${category}`);
  }

  async createProduct(data: Partial<Product>): Promise<Product> {
    return this.post<Product>(this.endpoint, data);
  }

  async updateProduct(id: string, data: Partial<Product>): Promise<Product> {
    return this.put<Product>(`${this.endpoint}/${id}`, data);
  }

  async deleteProduct(id: string): Promise<boolean> {
    await this.delete(`${this.endpoint}/${id}`);
    return true;
  }

  async getListingOptimization(productId: string): Promise<ProductListingOptimization> {
    return this.get<ProductListingOptimization>(`${this.endpoint}/${productId}/optimize`);
  }

  async incrementViewCount(productId: string): Promise<void> {
    await this.post(`${this.endpoint}/${productId}/view`);
  }
}
